### DISCUSSION TITLE
How to interact with a running python script in pyodide ?

### DISCUSSION THREAD

**Discussion by gabrielfougeron** [pyodide/pyodide#3259] [Q&A] [answered]
I am currently designing a small scientific computing app in python, including a web GUI thanks to pyodide. If that interests anyone, the project is in beta, and is available at [https://gabrielfougeron.github.io/choreo/](https://gabrielfougeron.github.io/choreo/). I've gotten very valuable help from the Pyodide dev team in the past, whom I thank.

In the GUI, the user can select parameters for the computation and then launch said computation (in pyodide). In order for the GUI to not freeze during the (sometimes a bit intensive) computation, pyodide is launched in a separate webworker.

Till now, the only way I found to interrupt a running computation is to kill the webworker and reload pyodide in a separate webworker.
If I post a message to this worker while Pyodide is running, the message isput on hold and only treated when Pyodide returns.

Is there a better way to interrupt a running computation in pyodide ?
How can I (for instance) deal with an event sent from the main Javascript directly in Pyodide?

Thank you very much

**Comment by hoodmane** [comment]
Look at`setInterruptBuffer`. I think there is some amount of tutorial info in the docs. (I can elaborate more later, posting on my phone.)

**Comment by gabrielfougeron** [comment]
Thanks for the pointer! I'll have a look and ask again fi I can't figure it out on my own.

**Comment by gabrielfougeron** [comment]
Looking at the example, if I understand correctly, this kind of interaction boils down to setting up a shared memory that can be accessed by both the worker and the main program, using a `SharedArrayBuffer`

However, this gives me the following error : `Uncaught ReferenceError: SharedArrayBuffer is not defined`

A quick search leaves me to believe this has to do with the CORS policy, but this is more or less how far I managed to go.

This leads me to believe there are security concerns associated. Is this true ?

**Comment by gabrielfougeron** [comment, marked as answer]
After further investigation, `SharedArrayBuffer `usage is indeed restricted by CORS. To serve files locally with the correct policy, I only slightly modified the `http_server.py` of the official pyodide docs as folllows:

```
import sys
import socketserver
from http.server import SimpleHTTPRequestHandler


class Handler(SimpleHTTPRequestHandler):

    def end_headers(self):

        # Enable Cross-Origin Resource Sharing (CORS)
        self.send_header('Access-Control-Allow-Origin', '*')

        self.send_header('Cross-Origin-Embedder-Policy', 'require-corp')
        self.send_header('Cross-Origin-Opener-Policy', 'same-origin')

        super().end_headers()


if sys.version_info < (3, 7, 5):
    # Fix for WASM MIME type for older Python versions
    Handler.extensions_map['.wasm'] = 'application/wasm'


if __name__ == '__main__':
    port = 8000
    with socketserver.TCPServer(("", port), Handler) as httpd:
        print("Serving at: http://127.0.0.1:{}".format(port))
        httpd.serve_forever()
```

**Reply by rth** [reply]
Thanks for sharing the solution!

**Comment by gabrielfougeron** [comment]
A last note:
For those (like me) who deploy their websites ont platforms where the HTTP headers cannot be tweaked,  [this github repo](https://github.com/gzuidhof/coi-serviceworker) lets you enable Cross-origin isolation through a service worker.
